### PR TITLE
Fixing Time Stamp Issue

### PR DESCRIPTION
**Pull request by john-b-yang**
The goal of this pull request is to replace the "time" field in the PelicanStatus struct with the time stamp of the most recent thermostat status update from the Pelican Web API. This way, users can tell if a Pelican thermostat is disconnected or communicating antiquated information.

The programmatic changes are listed as follows:
- Added “Location” field to params.yml file
- Added “Location” field to Pelican struct
- Added “Location” parameter to “NewPelican” and “DiscoverPelicans” functions and adjusted implementation slightly to create Pelican structs with location fields
- Added 3 new structs in pelican.go (apiResultHistory, apiRecords, apiHistory) to handle results of ThermostatHistory get request.
- Edited main.go so Spawnpoint requests user location and passes location as a parameter into DiscoverPelican function call
- Changed PelicanStatus struct’s “time” field to be a string, not int32 type
- Commenting to clarify what any new code is doing

### CODE REVIEW

File: driver/pelican/pelican.go
Comment (gtfierro) [inline driver/pelican/pelican.go]: Could we keep this as a 64-bit int timestamp as nanoseconds since the Unix Epoch? It would keep the interface consistent, and should be a straightforward change

File: driver/pelican/pelican.go
Comment (gtfierro) [inline driver/pelican/pelican.go]: It looks like here we're using only the timestamp of the historical record, but are reading "current status" from the thermostat API. How sure are we that these are going to be consistent? Would we want to pull the other thermostat state from the historical record as well? @john-b-yang you probably know this API pretty well by now, so I'll defer to you on what the best way to approach this is.

File: driver/pelican/params.yml
Comment (gtfierro) [inline driver/pelican/params.yml]: Can we put a comment above this saying that it must be a recognized IANA timezone name?

File: driver/pelican/main.go
Comment (jhkolb) [inline driver/pelican/main.go]: I know the underlying go type we're working with is called `Location`, but can we call this `timezone` instead? I think that will be clearer to the end user.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Let's call this `timezone` as well, since we're expecting it to be an IANA-compliant name for a timezone.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Drop the extra line here.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: And let's drop the extra line here as well.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: If I'm not mistaken, doesn't this need to be an `int64` expressing Unix time in nanoseconds to be compliant with the standard XBOS interfaces?

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Let's make this error message clearer. It's actually the result of an invalid timezone specified in the configuration, right?

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Let's change this so we parse the result and convert that to Unix time in nanoseconds. This will keep the thermostat interface consistent with the previous version.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: I think we can make this clearer. `diff` would probably be expected to represent the difference in time differences, not a `time.Time` instance that actually represents the start time.

I think it's okay to create two `time.Time` instances representing the start and end times, then just format them below in your use of `Param`.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Drop the extra line.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Hmm, we have two options here. One is to return an error if we can't find any sufficiently recent data, the other is to report the most recent status information we know about, even if it's old. @gtfierro what are your thoughts here?

My sense is that, if we look at this purely as a driver, then we'd always want to report the most recent status information, even if old. However, given that this sits in a larger ecosystem where we're publishing persisted messages on Bosswave, someone could just query the URI for this I guess. And I suppose this would generate nuisance traffic for things upstream like the archiver.

If we return an error when the data is too old, let's confirm that we're handling this properly in `main.go`. If we go with the always reporting option, here are some thoughts I sketched out on how to do this efficiently:

There are a few issues with the current approach:
  * What if the most recent timestamp is from more than 4 hours ago? And if we pick any fixed window size to look backwards, it's always possible that we'll miss data just outside of that window.
* We're only really using this API endpoint to find the most recent timestamp. We don't care about the other entries in the thermostat history, so we'll likely fetch unnecessary data.
* If we want to optimize for the common case where the thermostat is working properly, we don't need to look very far into the past to find the most recent timestamp.

Instead, let's use a loop where we first try looking 5 minutes (or however long the typical Pelican reporting period is) into the past. If we don't find anything, we look 10 minutes back, then 20 minutes, 40 minutes, and so on. Once we get actual results, we're done.

In fact, we could take this a step further. If, the last time we called `GetStatus`, we had to go 4 hours into the past to find the most recent timestamp, then on this call to the function, we could use that as our starting point. If we find a reading from 5 minutes ago, that'd be the window size for the next call to the function, etc. So the time window would grow and shrink based on past behavior, meaning we could usually get the result we want in just one iteration of the loop.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Let's just make this the idiomatic `== 0`.

**Review by jhkolb — CHANGES_REQUESTED** [review]
Nice, this is a good start! Some comments on suggested changes below, but all of the main pieces are in place.

File: driver/pelican/pelican.go
Comment (gtfierro) [reply driver/pelican/pelican.go]: I think the principle we want to work with is if the thermostat is offline, then we don't want to give the impression that it is online. This should manifest itself in 2 ways:
- stop publishing data once it is "old" (e.g. more than 1 minute old)
- stop publishing liveness messages for this interface; this will discourage attempted actuation of the device

Comment (jhkolb) [reply driver/pelican/pelican.go]: Ahh okay, we can forget my long blurb then. Lol.

What do you think is the best way to stop publishing liveness messages? I assume you're referring to the lastalive metadata stuff? My understanding is that would require unsubscribing from the slots...? If so, this is possible with some recent changes I made in `bw2bind` that were merged in.

**Comment by john-b-yang** [comment]
Hey @gtfierro @jhkolb! I've finished the edits you guys mentioned above.

First, I edited the naming convention of "location" and changed it to "timezone" in both the params.yml file and the pelican.go file. I left a comment in the "params.yml" file mentioning that the timezone should be IANA format. 

To address the problem regarding outdated statuses from the Pelican Web API, I decided that if no time stamps were found in the past 1 hour, the pelican's "GetStatus" method would return an error indicating that the web API may have stopped updating.

@gtfierro So to address the potential issue about the Time Stamp and Get Status struct's information not matching up, I ran a couple tests comparing the API results from the Thermostat History Object and the Thermostat Object. I haven't found any discrepancies between the values, so as of now, I don't think there's any reason to believe the TimeStamp and status might be out of sync. It's definitely a reasonable concern, but at the moment, the two seem to match up. I left the code as is for now, but if a problem arises, I can look into it.

Last but not least, I also did some formatting cleanup. Thanks @jhkolb for showing me the go-plus Atom library! Very helpful 👍

Let me know what you think!

File: driver/pelican/main.go
Comment (jhkolb) [inline driver/pelican/main.go]: I think you mean "sufficiently recent" data.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Let's remove the commented out line.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Are you sure you want to add the `00Z` suffix? That would imply that the timestamp is in UTC, but we know that Pelicans use their notion of local time.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: `timeInt` seems like it could mean a time interval but this is a timestamp, or value with an integer type but this is of type `time.Time`. So let's use a clearer name for this.

File: driver/pelican/pelican.go
Comment (john-b-yang) [reply driver/pelican/pelican.go]: That's a good point. The issue i'm running into right now is that if I don't add the suffix, the time.Parse() function will throw a parsing error at me. In other words, the timestamp that the pelican returns doesn't include the seconds and timezone, so for me to utilize the time.Parse function, I have to concatenate the :00Z suffix.

I guess one approach I could take is hard-coding a hashmap linking IANA timezones with their respective +/- in hours relative to UTC. Therefore, when I construct the timeString variable, I could append the correct timezone suffix.

I think the big problem with that approach is that the map will be extremely large due to all of the IANA timezones out there. I will look into whether there's some API I can use instead of the map. However, this is the only alternative I could conjure off the top of my head. I'm very open to suggestions.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go:228]: Didn't we say the reporting period for a thermostat was about 10 minutes? If so, I would think we only need to look about 20 minutes into the past.

The farther back we look, the longer the driver will wait before considering the thermostat unreachable and stopping status publication.

@gtfierro Any thoughts on this?

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Let's just call the instance of `time.Time` `timestamp`.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Ah I see, RFC 3339 is a stricter subset of ISO 8601 that requires the timezone to be specified.

Okay, I think you'll need to pass a custom format string to `time.Parse` instead of relying on the built-in `time.RFC3339`.

In fact, you can even use `time.ParseInLocation` since you'll know the thermostat's timezone from the params.

File: driver/pelican/pelican.go
Comment (john-b-yang) [reply driver/pelican/pelican.go:228]: Just to offer some extra insight, I just wrote a couple lines of Go that print out the list of timestamps from the past five hours, retrieved from the Thermostat History Object response. I ran this code at 10:30 p.m. From the results published below, I saw that the last push of data was at 21:52, or 9:52 p.m., and it seems that up until then, the thermostat has been updating in around 5-6 minute intervals.

I've seen this behavior before, where there seems to be a thirty minute gap between the current time and the last time stamp. I'm not sure what is accounting for this delay, but my hunch is that it's some latency from the pelican's information getting stored into the web api server. That's why I feel like an hour would be a more appropriate suggestion.

I included a screenshot below here, where I just ran the test at 11:45 p.m. Again, the last push is 11:17 p.m., which is why I feel like the hour time frame works pretty well. However, I'm open to suggestions 👍 

screen shot 2018-04-04 at 11 53 33 pm

File: driver/pelican/pelican.go
Comment (john-b-yang) [reply driver/pelican/pelican.go]: 👍

File: driver/pelican/pelican.go
Comment (jhkolb) [reply driver/pelican/pelican.go:228]: Okay, I think we're fine with 1 hour then. Just wanted to be sure we could justify our decision on how far back to look.

File: driver/pelican/params.yml
Comment (jhkolb) [inline driver/pelican/params.yml]: Hmm, I think we can leave the comment out here actually.

The other fields can have an expected syntax as well, e.g. `poll_interval` needs to be a valid Go-style time duration string, but we don't have comments for them, so let's just be consistent.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: This is still confusing because `diff` isn't a time difference, as its name would imply. It's actually a specific time. I say just make a `time.Time` instance from `time.Now` as a variable, then create `endTime` and `startTime` from it directly.

File: driver/pelican/params.yml
Comment (jhkolb) [reply driver/pelican/params.yml]: Nvm, sounds like we want to leave this in.

**Review by jhkolb — APPROVED** [review]

**Comment by jhkolb** [comment]
Looks good to me, so I went ahead and approved the PR.

Let's wait to see if @gtfierro wants to weigh in on anything before merging.

File: driver/pelican/pelican.go
Comment (gtfierro) [reply driver/pelican/pelican.go:228]: Did we double check what time the thermostat thinks it is? The time synchronization may be off. In any case, 1 hour does sound fine

**Comment by gtfierro** [comment]
I just built it and ran it and it looks to operate ok so far. How much testing have we done on this? @john-b-yang , can you deploy an instance of this driver against the pelican you've been using for testing and keep an eye on it for a couple days? You should disconnect the thermostat from power to see how the driver and pelican gateway handle the thermostat falling offline and see if the reporting matches what we expect. Once we've confirmed that this works as expected, then I say lets merge it. Great work @john-b-yang !

**Comment by john-b-yang** [comment]
Hey @jhkolb @gtfierro just a quick update! I rewrote the timestamp logic so that instead of requiring a user to manually enter their timezone, I found the timezone from the site's configuration. Consequently, I've removed the "timezone" field from the params.yml file and replaced all timezone fields with the site's built in timezone instead.

In addition, during this past week, I've tested the code by turning the Pelican on and off for different lengths of time. Everything seems to be in working order, and I don't believe there are any bugs or issues we still need to fix. Let me know if there's anything else I can do!

File: driver/pelican/main.go
Comment (jhkolb) [inline driver/pelican/main.go]: I don't think we want this logic in `main.go`. Rather, we should keep all interaction with Pelican's API within `pelican.go`.

Seems like you have two choices:
  1. Expose a `GetTimezone` function from `pelican.go` that can be called from `main.go`
  2. Fold the timezone discovery logic into the `DiscoverPelicans` function, so this handles figuring out the site's timezone as well as finding the pelican units present at that site.

**Review by jhkolb — CHANGES_REQUESTED** [review]

File: driver/pelican/main.go
Comment (jhkolb) [inline driver/pelican/main.go]: Remove these from `main.go`. All of the XML logic and any types for parsing XML should be fully encapsulated in `pelican.go`. The idea is that `main.go` doesn't care about how the Pelican API works, it just uses the data we get from the API and deals with Bosswave stuff.

File: driver/pelican/main.go
Comment (jhkolb) [inline driver/pelican/main.go]: Let's fold the timezone logic into the `DiscoverPelicans` function.

So we don't provide the timezone as an argument to `DiscoverPelicans`, instead it makes the API call to figure out the site's timezone and populates the pelican structs that are returned with that value.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: Let's make the type of this field `*time.Location` and use `time.LoadLocation` within the `DiscoverPelicans` logic. That way, we're not trying to parse the timezone descriptor every time we retrieve the thermostat's status.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: `Object` is pretty vague. The type of the object you're actually retrieving here is called `Site` by the Pelican documentation, just like we had `Thermostat` and `ThermostatHistory` before.

You should update the comment and type names so they're more clear.

File: driver/pelican/pelican.go
Comment (jhkolb) [inline driver/pelican/pelican.go]: This logic will move into `DiscoverPelicans`, and now you can do something along the lines of `time.Now().In(pel.timezone)` to construct the start and end timestamps.

**Review by jhkolb — APPROVED** [review]

**Comment by jhkolb** [comment]
@gtfierro this has been pretty well tested over the past couple of weeks, so if you're content with the changes to the code, I think this is ready to be merged.

**Comment by gtfierro** [comment]
Yay! Lets merge!

**Comment by gtfierro** [comment]
The reason I reverted this PR is because it introduces the behavior of only reading the latest historical point (as recorded by the Pelican API). The history only stores the *changes* in temperature. Right now, we associate the timestamp of reported data with liveness, so when the thermostat driver only publishes the changes in temperature, we get duplicate timestamps for however long the temperature is the same. This can result in large time gaps in the database, which confuses most of the analytics and alarms we've built, which assume that data is more or less continuously flowing, and an absence of data readings means that the driver has crashed or device has gone offline.

This does mean that we're in a bit of a spot with how to distinguish between a stable (but live) thermostat and a dead thermostat. I'll investigate this with the Pelican folks and update here what they recommend. For now I'm reverting this change so that alarms/analysis continue to work

**Pull request merged**
